### PR TITLE
fix bug in maxout function

### PR DESCRIPTION
**Pull request by lygztq**
The line "shape[axis] = -1" will make the shape wrong when dealing with batches with arbitrary sizes.
if the shape of input tensor is [None, ... , num_channels, ... ], "shape[axis]=-1" together with "shape+=[num_channels // num_units]" will make the shape become [None, ... , -1, ... , num_channels // num_units]. But when reshape the input tensor, the "None" element in shape will make "-1" become "None", not "num_units".

### CODE REVIEW

**Comment by av8ramit** [comment]
@martinwicke can I merge this?

**Comment by martinwicke** [comment]
Check with @yifeif but I think yes.

**Comment by yifeif** [comment]
You are good if it's for non-master branches.

**Comment by av8ramit** [comment]
MacOS flake. Re-running.

File: tensorflow/contrib/layers/python/layers/layers.py [hunk 1]
```diff
@@ -3117,7 +3117,7 @@ def maxout(inputs, num_units, axis=-1, scope=None):
       raise ValueError('number of features({}) is not '
                        'a multiple of num_units({})'.format(
                            num_channels, num_units))
-    shape[axis] = -1
+    shape[axis] = num_units
```
Comment (pavithrasv) [inline tensorflow/contrib/layers/python/layers/layers.py]: Can we add a test case with `None` value for an axis?

**Review by pavithrasv — APPROVED** [review]

**Comment by case540** [comment]
Going to merge this fix into 1.9 so we can start official release builds. Please follow-up with @pavithrasv comment and add a test case with None please though.

Thanks!

File: tensorflow/contrib/layers/python/layers/layers.py (hunk 1 above)
Comment (lygztq) [reply tensorflow/contrib/layers/python/layers/layers.py]: Like this? I am not sure.
```python
input_tensor =tf.placeholder(tf.float32, [None, 32, 32, 3], name="input_data")
with tf.variable_scope('conv_layer'):
    conv_out = tf.layers.conv2d(input_tensor, 10, (5,5), padding='same', name="conv")
    maxout_out = tf.contrib.layers.maxout(conv_out_1, 5, axis=3)
#... some test code.
self.assertEqual(maxout_out.get_shape().as_list()[3], 5, "shape in maxout function failed")
```

Comment (lygztq) [reply tensorflow/contrib/layers/python/layers/layers.py]:
```python
import numpy as np
import tensorflow as tf

# Suppose this function belong to some test class
def testMaxoutShape(self):
    data = np.random.randn(64, 32, 32, 3).astype(np.float32)
    input_tensor = tf.placeholder(tf.float32, [None, 32, 32, 3], name="input_data")
    conv_out = tf.layers.conv2d(input_tensor, 10, (5,5), padding='same', name="conv")
    maxout_out = tf.contrib.layers.maxout(conv_out, 5, axis=3)

    with tf.Session() as sess:
        sess.run(tf.global_variables_initializer())
        value = sess.run(maxout_out, feed_dict={input_tensor: data})
    self.assertEqual(value.shape[3], 5, "shape in maxout function failed")
```

Comment (pavithrasv) [reply tensorflow/contrib/layers/python/layers/layers.py]: Would you like to make the change to the test file? https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/python/layers/layers_test.py#L4172
